File: datasets/Python/Deepseek/DataStructures/medium4.py
class TrieNode:
    def __init__(self):
        self.children = {}
        self.count = 0

class Contacts:
    def __init__(self):
        self.root = TrieNode()

    def add(self, name):
        node = self.root
        for char in name:
            if char not in node.children:
                node.children[char] = TrieNode()
            node = node.children[char]
            node.count += 1

    def find(self, partial):
        node = self.root
        for char in partial:
            if char not in node.children:
                return 0
            node = node.children[char]
        return node.count

def contacts(queries):
    contact_list = Contacts()
    results = []
    for query in queries:
        parts = query.split()
        if parts[0] == 'add':
            contact_list.add(parts[1])
        elif parts[0] == 'find':
            results.append(contact_list.find(parts[1]))
    return results

if __name__ == '__main__':
    n = int(input())
    queries = [input().strip() for _ in range(n)]
    output = contacts(queries)
    for result in output:
        print(result)

File: datasets/C++/Deepseek/DataStructures/medium4.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
#include <string>

using namespace std;

class TrieNode {
public:
    unordered_map<char, TrieNode*> children;
    int count = 0;
};

class Contacts {
private:
    TrieNode* root;

public:
    Contacts() {
        root = new TrieNode();
    }

    void add(const string& name) {
        TrieNode* node = root;
        for (char c : name) {
            if (node->children.find(c) == node->children.end()) {
                node->children[c] = new TrieNode();
            }
            node = node->children[c];
            node->count++;
        }
    }

    int find(const string& partial) {
        TrieNode* node = root;
        for (char c : partial) {
            if (node->children.find(c) == node->children.end()) {
                return 0;
            }
            node = node->children[c];
        }
        return node->count;
    }
};

vector<int> contacts(const vector<string>& queries) {
    Contacts contactList;
    vector<int> results;
    for (const string& query : queries) {
        size_t space_pos = query.find(' ');
        string cmd = query.substr(0, space_pos);
        string arg = query.substr(space_pos + 1);
        if (cmd == "add") {
            contactList.add(arg);
        } else if (cmd == "find") {
            results.push_back(contactList.find(arg));
        }
    }
    return results;
}

int main() {
    int n;
    cin >> n;
    cin.ignore();
    vector<string> queries(n);
    for (int i = 0; i < n; ++i) {
        getline(cin, queries[i]);
    }
    vector<int> output = contacts(queries);
    for (int result : output) {
        cout << result << endl;
    }
    return 0;
}